include/dlaf/communication/kernels/all_reduce.h

```diff
@@ -89,52 +72,60 @@ void scheduleAllReduce(CommSender&& pcomm, MPI_Op reduce_op,
-    auto tile_reduced =
-        whenAllLift(std::move(pcomm), reduce_op, makeItContiguous(tile_in), makeItContiguous(tile_out)) |
-        transformMPI(internal::allReduce_o);
-    return whenAllLift(std::move(tile_reduced), std::cref(tile_out)) |
-           transform(Policy<Backend::MC>(thread_priority::high), copyBack_o);
-  });
-  ex::when_all(ex::keep_future(std::move(tile_in)), std::move(tile_out)) |
-      ex::transfer(getBackendScheduler<Backend::MC>()) | ex::let_value(std::move(f)) |
-      ex::start_detached();
-}
-
-template <class CommSender, class TSender>
-[[nodiscard]] auto scheduleAllReduceInPlace(CommSender&& pcomm, MPI_Op reduce_op, TSender&& tile) {
-  namespace ex = pika::execution::experimental;
-
-  using common::internal::copyBack_o;
-  using common::internal::makeItContiguous;
-  using dlaf::internal::getBackendScheduler;
-  using dlaf::internal::Policy;
-  using dlaf::internal::transform;
+  using dlaf::comm::internal::allReduce_o;
+  using dlaf::comm::internal::transformMPI;
+  using dlaf::internal::CopyFromDestination;
+  using dlaf::internal::CopyToDestination;
+  using dlaf::internal::RequireContiguous;
   using dlaf::internal::whenAllLift;
-  using pika::threads::thread_priority;
-
-  using T = dlaf::internal::SenderElementType<TSender>;
+  using dlaf::internal::withTemporaryTile;
+
+  auto all_reduce_final = [reduce_op, pcomm = std::forward<CommSender>(pcomm),
+                           tile_in =
+                               std::forward<TileInSender>(tile_in)](auto const& tile_out_comm) mutable {
+    auto all_reduce = [reduce_op, pcomm = std::move(pcomm),
+                       &tile_out_comm](auto const& tile_in_comm) mutable {
+      return whenAllLift(std::move(pcomm), reduce_op, std::cref(tile_in_comm),
+                         std::cref(tile_out_comm)) |
+             transformMPI(allReduce_o);
+    };
+    return withTemporaryTile<Device::CPU, CopyToDestination::Yes, CopyFromDestination::No,
+                             RequireContiguous::Yes>(std::move(tile_in), std::move(all_reduce));
+  };
+
+  // TODO: Can reductions happen on CPU only?
+  return withTemporaryTile<Device::CPU, CopyToDestination::No, CopyFromDestination::Yes,
+                           RequireContiguous::Yes>(std::forward<TileOutSender>(tile_out),
+                                                   std::move(all_reduce_final));
+}
 
+template <class CommSender, class TileSender>
+[[nodiscard]] auto scheduleAllReduceInPlace(CommSender&& pcomm, MPI_Op reduce_op, TileSender&& tile) {
   // Note:
   //
   // TILE ---> makeContiguous --+--> CONT_BUF ----> mpi_call ---> CONT_BUF --+
   //                            |                                            |
   //                            +------------------> TILE -------------------+-> copyBack ---> TILE
   //
-  //
   // The last TILE after the copyBack is returned so that other task can be attached to it,
   // AFTER the asynchronous MPI_AllReduce has completed
```

Comment:
How would you prefer to see these comments updated?

My preference would be to change the comments in the `scheduleX` functions to explain why the template parameters are as they are in the call to `withTemporaryTile` and remove the diagrams in `scheduleX`. I would instead add one or more diagrams to the `withTemporaryTile` documentation instead explaining the general idea of creating a temporary tile. :+1: if you're happy with this or comment if you prefer something else.

Reply:
I've updated this. I'm marking this as resolved, but please feel free to reopen/comment if you'd like to see changes.